Sai Venkatesh - frontend - Improve speed and function of deleting multiple blue squares

SFA23SCM35V opened the pull request:
# Description
Blue squares used to come back, now they just delete really slowly or not at all.

## Related PRS :
This frontend PR is related to the #[1130](https://github.com/OneCommunityGlobal/HGNRest/pull/1130) backend PR.

…

## Main changes explained:
Consolidated destructuring for props.
Simplified handleOnClick by reducing repeated conditions.
Moved formatting logic inside JSX.
Removed redundant conditions and comments for cleaner code.
…

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as Owner user
5. Profile>Blue Square Section>Delete a bunch in a row
6. verify functionality 

https://www.loom.com/share/02a9fd342bf84425abdf547566c85035

chu-ziyu reviewed (commented):
Hi Sai,
I really appreciated your work but when I was trying to test this PR I have trouble fetching your branch. It shows as the screenshot below, I thought I am not the only one who met this problem. Feel free to reach out to me. Thank you.
Ziyu Chu
![Screenshot 2024-10-11 at 6 36 42 PM](https://github.com/user-attachments/assets/af855aea-fa41-49e4-ad3f-8bc42d3d98a9)

nathanah commented on src/components/UserProfile/BlueSquares/BlueSquare.jsx:
```diff
@@ -3,77 +3,151 @@ import hasPermission from 'utils/permissions';
+//                   <div className="report" data-testid="report">
+//                     <div className="title">{formatDate(blueSquare.date)}</div>
+//                     {blueSquare.description !== undefined && (
+//                       <div className="summary">
+//                         {blueSquare.createdDate !== undefined && blueSquare.createdDate !== null
+//                           ? `${formatCreatedDate(blueSquare.createdDate)}: ${
+//                               blueSquare.description
+//                             }`
+//                           : blueSquare.description}
+//                       </div>
+//                     )}
+//                   </div>
+//                 </div>
+//               ))
+//           : <div>No blue squares.</div>}
+//         {canAddInfringements && (
+//           <div
+//             onClick={() => {
+//               handleBlueSquare(true, 'addBlueSquare', '');
+//             }}
+//             className="blueSquareButton"
+//             color="primary"
+//             data-testid="addBlueSquare"
+//           >
+//             +
+//           </div>
+//         )}
+//       </div>
+//     </div>
+//   );
+// };
+
+
+/*
+Consolidated destructuring for props.
+Simplified handleOnClick by reducing repeated conditions.
+Moved formatting logic inside JSX.
+Removed redundant conditions and comments for cleaner code.
+*/
+
```
```suggestion
```
Same as the backend. Deleting this old code instead of just commenting it out and duplicating it reduces clutter from old code. The comment should be in the commit/PR description as it doesn't make sense to keep after the PR.

nathanah reviewed (changes requested):
Same thing as backend. Clean up the comments and I'll approve. The refactor seems fine, but I'm not sure how exactly it "Improve[s] speed and function" as the title indicates.

ishangoel17 reviewed (commented):
Unable to review the bug as I am getting error when trying to fetch the branch. Below is the screenshot for reference:
![Capture2774](https://github.com/user-attachments/assets/77433fec-d000-4038-a5a8-30b07d989e5f)

CodewithKoushica commented:
Hello Sai,
I went through the code changes and I really like how organized your code is—great job!
However, I can’t find the branch for this pull request. Can you please confirm that it’s pushed to the remote repository?
I see you’re fixing the issue with Blue Squares not deleting properly, and I’d like to check the performance once I have access to the code.

I appreciate your work on this!

<img width="808" alt="Screenshot 2024-10-15 at 11 21 18 AM" src="https://github.com/user-attachments/assets/9419683a-01c0-4b93-a47b-8d2b3c73ebdf">

Ankuriboh reviewed (changes requested):
Simple refactor, should be ready to merge. Before that, please kindly remove all unnecessary comments from the code. 

Also, for further improvements, please kindly create a new branch in this repo instead of branching out in a forked repo.

sheetalmangate commented:
Hello Sai,

The code refactoring appears to be correct, but the commented code should be removed.  As per the task heading 'Improve speed and function...,' there doesn’t seem to be any code optimization present to enhance performance.

Additionally while testing locally, I discovered an issue :: The blue square view is not loading correctly for the newly added blue square before refreshing the page. This problem also exists on development branch.

SFA23SCM35V commented:
@sheetalmangate 
I have removed the comments. Please raise a new bug in the bug sheet, as this PR is not related to the view component. Could you also verify whether the blue square is being deleted as expected? If everything works fine, please approve the PR

SFA23SCM35V commented:
@Ankuriboh 
removed the comments and will create a new branch in this repo from now 
if everything works well please Approve the PR.

SFA23SCM35V commented:
@ishangoel17 
try synchronize changes before checking to the branch get the latest repo

SFA23SCM35V commented:
@nathanah 

1.  Consolidated Destructuring for Props
By destructuring props at the start of the function:
const { blueSquares, handleBlueSquare, hasPermission } = props;

Benefit: This reduces redundant lookups like props.hasPermission, making the code more concise and slightly faster at runtime (though the performance gain is negligible in modern JavaScript engines). The main benefit here is readability and easier debugging, which reduces development time when modifying or maintaining the delete functionality.

2. Simplified handleOnClick by Reducing Repeated Conditions
Before: You had multiple conditions handling edit, delete, and view actions separately.
After: You combined similar conditions into a single call, like so:

if (canEditInfringements || canDeleteInfringements) {
  handleBlueSquare(true, 'modBlueSquare', blueSquare._id);
} else {
  handleBlueSquare(true, 'viewBlueSquare', blueSquare._id);
}

Benefit: This makes the logic easier to follow, reduces unnecessary branches, and ensures quicker execution when deleting. Although the performance improvement is minimal, cleaner logic reduces the chance of bugs, making the delete functionality more reliable and easier to optimize in the future.

4. Moved Formatting Logic Inside JSX
Before: You were checking for conditions and formatting strings in separate blocks.
After: You moved this into the JSX for a more streamlined rendering process:

{blueSquare.createdDate ? `${formatCreatedDate(blueSquare.createdDate)}: ` : ''}
{blueSquare.description}

Benefit: This minimizes unnecessary intermediate logic and lets React handle the rendering more efficiently. This doesn’t directly speed up the deletion but makes rendering the component more efficient, especially in complex UIs with lots of elements. By reducing unnecessary overhead in rendering, React can potentially free up resources faster when performing state updates like deletion.

5. Removed Redundant Conditions and Comments
Benefit: Removing unused or redundant conditions reduces the number of checks the system has to make when rendering or handling click events. While this has minimal direct performance impact, it improves the overall flow of the code, making it easier to read, debug, and optimize in the future.

How These Changes Affect Deletion Speed
Reduced Logical Overhead: Simplifying the conditions for clicking (i.e., when a user clicks to delete a blue square) means the event handler is more straightforward, with fewer branching conditions. This can slightly reduce the time it takes for the function to determine what action to perform (edit, delete, or view), making deletion a bit faster in terms of execution.

Better Maintainability: Cleaner code is easier to modify and optimize. If you ever need to further optimize the delete functionality (e.g., through better state management or using React.memo to prevent unnecessary re-renders), having a cleaner base will make these optimizations more effective and easier to implement.

If you still feel it doesn't improve the speed, feel free to suggest changes.

SFA23SCM35V commented:
@chu-ziyu 

try synchronize changes before checking to the branch get the latest repo

Rui-Liu-github reviewed (commented):
Hi, I reviewed a PR #2774, It's good to optimize the delete functionality. Optimizing the code can reduce the frequency of future modifications by other developers and improve readability.   If you want to optimize the delete function, would it be possible to select the chosen squares and delete it with a just single click?   Currently, it is possible to delete individual squares consecutively.

Rui-Liu-github reviewed (approved):
HI, I reviewed the PR #2774 , I checked the delete multiple blue squares functionality which could be done well.

samarth9201 reviewed (commented):
The code looks good, however unable to test it as I am not able to fetch the branch.

![image](https://github.com/user-attachments/assets/4ef98319-d5b9-42d7-99d4-b34b3e65bf0f)

Rupa2400 reviewed (approved):
PR #2774 I reviewed the functionality for deleting multiple blue squares. The blue squares are successfully removed from the profile, and they do not reappear after refreshing the page. The changes appear to be functioning as expected and align with the intended behavior.


https://github.com/user-attachments/assets/85edbd38-906c-433f-8316-f4bedd1211b9

AshritaCherlapally commented:
The code changes looks great, but could not able to test the functionality as I am unable to check in to the current branch.

<img width="1159" alt="PR#2774" src="https://github.com/user-attachments/assets/e0ca9795-3f80-41d6-a712-61e71774932f">
